tests/30rooms/06invite.pl

```diff
@@ -107,3 +111,424 @@
+#
+#         matrix_create_room->( $inviter, visibility => "private" )
+#         ->then(sub {
+#            my ( $room_id ) = @_;
+#            do_3pid_invite( $room_id, $id_server, $invitee_email, $do_request_json )
+#            ->then( sub {
+#               do_request_json_for( $invitee,
+#                  method => "POST",
+#                  uri    => "/api/v1/rooms/$room_id/join",
+#                  content => {
+#                     token => $token,
+#                     public_key => encode_base64_unpadded( $public_key ),
+#                     signature => $signature,
+#                     key_validity_url => "https://$id_server/_matrix/identity/api/v1/pubkey/isvalid",
+#                     sender => $user->user_id,
+#                  }
+#               )
+#               ->followed_by(\&main::expect_http_4xx)
+#               ->then( sub {
+#                  $do_request_json->(
+#                     method => "GET",
+#                     uri    => "/api/v1/rooms/$room_id/state/m.room.member/".$invitee->user_id,
+#                  )->followed_by(assert_membership( $do_request_json, $inviter, undef ) )
+#               })
+#            })
+#         }),
+#      );
+#   };
+
+# In order:
+#  1. Creates all state needed to issue a 3pid invite
+#  2. Issues the invite
+#  3. Calls join_sub with the following args: token (str), public_key (base64 str), $signature (base64 str), room_id (str)
+#  4. Asserts that invitee did/didn't join the room, depending on truthiness of expect_join_success
+#  5. Awaits on all passed futures, so that you can stub/mock things as you wish
+sub make_3pid_invite {
+   my ( $inviter, $invitee, $id_server, $await_http_request, $expect_join_success, $join_sub, $futures ) = @_;
+
+   my $invitee_email = 'lemurs@monkeyworld.org';
+
+   my $token = "abc123";
+
+   my ( $public_key, $private_key ) = $crypto_sign->keypair;
+   my $encoded_public_key = encode_base64_unpadded( $public_key );
+   my $signature = encode_base64_unpadded( $crypto_sign->mac( $token, $private_key ) );
+
+   my $response_verifier = $expect_join_success
+      ? sub {
+         my ( $f ) = @_;
+         $f->then( sub {
+            Future->done( @_ );
+         }, sub {
+            my ( undef, $name, $response ) = @_;
+
+            Future->fail( @_ );
+         });
+      }
+      : \&main::expect_http_4xx;
+
+   Future->needs_all(
+      stub_is_lookup( $invitee_email, undef, $await_http_request ),
+
+      stub_is_token_generation( $token, $encoded_public_key, $await_http_request ),
+
+      @$futures,
+
+      matrix_create_room( $inviter, visibility => "private" )
+      ->then(sub {
+         my ( $room_id ) = @_;
+         do_3pid_invite( $inviter, $room_id, $id_server, $invitee_email )
+         ->then( sub {
+            $join_sub->( $token, $public_key, $signature, $room_id )
+            ->followed_by($response_verifier)
+            ->then( sub {
+               matrix_get_room_state( $inviter, $room_id,
```

Comment:
Most of this nesting doesn't really seem to be required. Furthermore, if you pulled out `my $room_id;` to above the `->needs_all` invocation, then the whole lot could become flat rather than pyramid-shaped, thus saving indentation levels.